instrumentation/opentelemetry-instrumentation-kafka-python/src/opentelemetry/instrumentation/kafka/utils.py

```diff
@@ -0,0 +1,215 @@
+            if item_key == key:
+                if value is not None:
+                    return [value.decode()]
+        return None
+
+    def keys(self, carrier: textmap.CarrierT) -> List[str]:
+        if carrier is None:
+            return []
+        return [key for (key, value) in carrier]
+
+
+class KafkaContextSetter(textmap.Setter):
+    def set(self, carrier: textmap.CarrierT, key: str, value: str) -> None:
+        if carrier is None or key is None:
+            return
+
+        if value:
+            value = value.encode()
+        carrier.append((key, value))
+
+
+_kafka_getter = KafkaContextGetter()
+_kafka_setter = KafkaContextSetter()
+
+
+def _enrich_span(
+    span, bootstrap_servers: List[str], topic: str, partition: int
+):
+    if span.is_recording():
+        span.set_attribute(SpanAttributes.MESSAGING_SYSTEM, "kafka")
+        span.set_attribute(SpanAttributes.MESSAGING_DESTINATION, topic)
+        span.set_attribute(SpanAttributes.MESSAGING_KAFKA_PARTITION, partition)
+        span.set_attribute(
+            SpanAttributes.MESSAGING_URL, json.dumps(bootstrap_servers)
+        )
+
+
+def _get_span_name(operation: str, topic: str):
+    return f"kafka.{operation}: {topic}"
+
+
+def _wrap_send(tracer: Tracer, produce_hook: HookT) -> Callable:
+    def _traced_send(func, instance, args, kwargs):
+        headers = KafkaPropertiesExtractor.extract_send_headers(args, kwargs)
+        if headers is None:
+            headers = []
+            kwargs["headers"] = headers
+
+        topic = KafkaPropertiesExtractor.extract_send_topic(args)
+        bootstrap_servers = KafkaPropertiesExtractor.extract_bootstrap_servers(
+            instance
+        )
+        partition = KafkaPropertiesExtractor.extract_send_partition(
+            instance, args, kwargs
+        )
+        span_name = _get_span_name("send", topic)
+        with tracer.start_as_current_span(
+            span_name, kind=trace.SpanKind.PRODUCER
+        ) as span:
+            _enrich_span(span, bootstrap_servers, topic, partition)
+            inject(
+                headers,
+                context=set_span_in_context(span),
+                setter=_kafka_setter,
+            )
+            try:
+                produce_hook(span, args, kwargs)
+            except Exception as hook_exception:  # pylint: disable=W0703
+                _LOG.exception(hook_exception)
+
+        return func(*args, **kwargs)
+
+    return _traced_send
+
+
+def _start_consume_span_with_extracted_context(
+    tracer: Tracer, headers: List, topic: str
+) -> Span:
+    extracted_context = extract(headers, getter=_kafka_getter)
+    span_name = _get_span_name("consume", topic)
+    span = tracer.start_span(
+        span_name, context=extracted_context, kind=trace.SpanKind.CONSUMER
+    )
+    new_context = set_span_in_context(span, extracted_context)
+    attach(new_context)
+    return span
+
+
+def _wrap_next(tracer: Tracer, consume_hook: HookT) -> Callable:
```

Comment:
This is a bit unusual. Will this work correctly with multiple consumers? What if user code raises an exception and _next__ is never called after creating a span? Will we end up never finishing the span and in case of a lot of exceptions result in a "memory leak"?

Reply:
the motivation behind doing that is having this `receive` span as a parent of every other span being created while processing the message.

but you are right, it isn't the best solution, I'll try to find another way to do that